File: demo-app/ios/ObserveSDK/Selectors/SelectorBuilder.swift
//
//  SelectorBuilder.swift
//  ObserveSDK
//
//  Enhanced selector builder for iOS native elements
//

import UIKit
import SwiftUI

/// Enhanced selector builder for iOS elements
/// Generates multiple selector strategies with uniqueness validation
public class SelectorBuilder {
    
    /// Element selector information with multiple strategies
    public struct ElementSelectors: Codable {
        public let accessibilityIdentifier: String?
        public let accessibilityLabel: String?
        public let text: String?
        public let className: String
        public let viewIdPath: String
        public let hierarchyPath: String
        public let indexedPath: String
        public let bounds: Bounds?
        public let attributes: [String: String]
        
        public struct Bounds: Codable {
            public let x: Double
            public let y: Double
            public let width: Double
            public let height: Double
        }
        
        /// Get primary selector (most reliable available)
        public func getPrimarySelector() -> String {
            if let accessibilityId = accessibilityIdentifier {
                return "accessibility_id:\(accessibilityId)"
            } else if let label = accessibilityLabel {
                return "accessibility_label:\(label)"
            } else if let text = text, text.count < 50 {
                return "text:\(text)"
            } else {
                return viewIdPath
            }
        }
        
        /// Get all selector strategies in priority order
        public func getAllStrategies() -> [(String, String)] {
            var strategies: [(String, String)] = []
            
            if let id = accessibilityIdentifier {
                strategies.append(("accessibilityIdentifier", "accessibility_id:\(id)"))
            }
            if let label = accessibilityLabel {
                strategies.append(("accessibilityLabel", "accessibility_label:\(label)"))
            }
            if let text = text, text.count < 50 {
                strategies.append(("text", "text:\(text)"))
            }
            strategies.append(("viewIdPath", viewIdPath))
            strategies.append(("hierarchyPath", hierarchyPath))
            strategies.append(("indexedPath", indexedPath))
            
            return strategies
        }
        
        /// Convert to JSON string
        public func toJSON() -> String? {
            let encoder = JSONEncoder()
            encoder.outputFormatting = .prettyPrinted
            guard let data = try? encoder.encode(self),
                  let json = String(data: data, encoding: .utf8) else {
                return nil
            }
            return json
        }
    }
    
    /// Build selectors for a UIView
    public static func buildSelectors(for view: UIView, in rootView: UIView) -> ElementSelectors {
        let accessibilityId = view.accessibilityIdentifier
        let accessibilityLabel = view.accessibilityLabel
        let text = extractText(from: view)
        let className = String(describing: type(of: view))
        
        let viewPath = buildViewPath(for: view, in: rootView)
        let hierarchyPath = buildHierarchyPath(for: view, in: rootView)
        let indexedPath = buildIndexedPath(for: view, in: rootView)
        
        let frame = view.frame
        let bounds = ElementSelectors.Bounds(
            x: Double(frame.origin.x),
            y: Double(frame.origin.y),
            width: Double(frame.size.width),
            height: Double(frame.size.height)
        )
        
        let attributes = extractAttributes(from: view)
        
        return ElementSelectors(
            accessibilityIdentifier: accessibilityId,
            accessibilityLabel: accessibilityLabel,
            text: text,
            className: className,
            viewIdPath: viewPath,
            hierarchyPath: hierarchyPath,
            indexedPath: indexedPath,
            bounds: bounds,
            attributes: attributes
        )
    }
    
    /// Extract text content from various view types
    private static func extractText(from view: UIView) -> String? {
        if let label = view as? UILabel {
            return label.text
        } else if let button = view as? UIButton {
            return button.titleLabel?.text ?? button.currentTitle
        } else if let textField = view as? UITextField {
            return textField.placeholder ?? textField.text
        } else if let textView = view as? UITextView {
            return textView.text
        }
        return nil
    }
    
    /// Build path using accessibility identifiers where available
    private static func buildViewPath(for view: UIView, in rootView: UIView) -> String {
        var path: [String] = []
        var current: UIView? = view
        
        while current != nil && current != rootView {
            guard let currentView = current else { break }
            
            let className = String(describing: type(of: currentView))
            let segment: String
            
            if let accessibilityId = currentView.accessibilityIdentifier {
                segment = "\(className)[@id='\(accessibilityId)']"
            } else if let label = currentView.accessibilityLabel {
                segment = "\(className)[@label='\(label)']"
            } else {
                segment = className
            }
            
            path.insert(segment, at: 0)
            current = currentView.superview
        }
        
        return "/" + path.joined(separator: "/")
    }
    
    /// Build path using class hierarchy
    private static func buildHierarchyPath(for view: UIView, in rootView: UIView) -> String {
        var path: [String] = []
        var current: UIView? = view
        
        while current != nil && current != rootView {
            guard let currentView = current else { break }
            let className = String(describing: type(of: currentView))
            path.insert(className, at: 0)
            current = currentView.superview
        }
        
        return "/" + path.joined(separator: "/")
    }
    
    /// Build indexed path (most fragile, but always works)
    private static func buildIndexedPath(for view: UIView, in rootView: UIView) -> String {
        var path: [String] = []
        var current: UIView? = view
        
        while current != nil && current != rootView {
            guard let currentView = current else { break }
            
            let className = String(describing: type(of: currentView))
            let index: Int
            
            if let superview = currentView.superview {
                index = superview.subviews.firstIndex(of: currentView) ?? 0
            } else {
                index = 0
            }
            
            path.insert("\(className)[\(index)]", at: 0)
            current = currentView.superview
        }
        
        return "/" + path.joined(separator: "/")
    }
    
    /// Extract all relevant attributes
    private static func extractAttributes(from view: UIView) -> [String: String] {
        var attrs: [String: String] = [:]
        
        attrs["isUserInteractionEnabled"] = String(view.isUserInteractionEnabled)
        attrs["isHidden"] = String(view.isHidden)
        attrs["alpha"] = String(view.alpha)
        attrs["isAccessibilityElement"] = String(view.isAccessibilityElement)
        
        // Type-specific attributes
        if let button = view as? UIButton {
            attrs["isEnabled"] = String(button.isEnabled)
            attrs["isSelected"] = String(button.isSelected)
        } else if let textField = view as? UITextField {
            if let placeholder = textField.placeholder {
                attrs["placeholder"] = placeholder
            }
            attrs["isSecureTextEntry"] = String(textField.isSecureTextEntry)
        } else if let switchControl = view as? UISwitch {
            attrs["isOn"] = String(switchControl.isOn)
        } else if let slider = view as? UISlider {
            attrs["value"] = String(slider.value)
            attrs["minimumValue"] = String(slider.minimumValue)
            attrs["maximumValue"] = String(slider.maximumValue)
        }
        
        return attrs
    }
    
    /// Validate if selector is unique in the view hierarchy
    public static func isSelectorUnique(_ selector: String, in rootView: UIView) -> Bool {
        let matchCount = countMatches(for: selector, in: rootView)
        return matchCount == 1
    }
    
    /// Count how many views match the given selector
    private static func countMatches(for selector: String, in rootView: UIView) -> Int {
        // Parse selector format
        if selector.hasPrefix("accessibility_id:") {
            let id = String(selector.dropFirst("accessibility_id:".count))
            return countByAccessibilityId(id, in: rootView)
        } else if selector.hasPrefix("accessibility_label:") {
            let label = String(selector.dropFirst("accessibility_label:".count))
            return countByAccessibilityLabel(label, in: rootView)
        } else if selector.hasPrefix("text:") {
            let text = String(selector.dropFirst("text:".count))
            return countByText(text, in: rootView)
        } else if selector.hasPrefix("/") {
            // Path-based selector (viewIdPath, hierarchyPath, or indexedPath)
            return countByPath(selector, in: rootView)
        } else {
            // Unknown format
            return 0
        }
    }
    
    /// Count views with matching accessibility identifier
    private static func countByAccessibilityId(_ id: String, in rootView: UIView) -> Int {
        var count = 0
        traverseHierarchy(rootView) { view in
            if view.accessibilityIdentifier == id {
                count += 1
            }
        }
        return count
    }
    
    /// Count views with matching accessibility label
    private static func countByAccessibilityLabel(_ label: String, in rootView: UIView) -> Int {
        var count = 0
        traverseHierarchy(rootView) { view in
            if view.accessibilityLabel == label {
                count += 1
            }
        }
        return count
    }
    
    /// Count views with matching text content
    private static func countByText(_ text: String, in rootView: UIView) -> Int {
        var count = 0
        traverseHierarchy(rootView) { view in
            if let viewText = extractText(from: view), viewText == text {
                count += 1
            }
        }
        return count
    }
    
    /// Count views matching a path-based selector
    private static func countByPath(_ path: String, in rootView: UIView) -> Int {
        var count = 0
        traverseHierarchy(rootView) { view in
            let viewIdPath = buildViewPath(for: view, in: rootView)
            let hierarchyPath = buildHierarchyPath(for: view, in: rootView)
            let indexedPath = buildIndexedPath(for: view, in: rootView)
            
            if viewIdPath == path || hierarchyPath == path || indexedPath == path {
                count += 1
            }
        }
        return count
    }
    
    /// Traverse view hierarchy and execute callback for each view
    private static func traverseHierarchy(_ view: UIView, callback: (UIView) -> Void) {
        callback(view)
        for subview in view.subviews {
            traverseHierarchy(subview, callback: callback)
        }
    }

}

// MARK: - SwiftUI Support

extension SelectorBuilder {
    
    /// Build selectors for SwiftUI views (when accessible via UIHostingController)
    public static func buildSelectorsForSwiftUI(identifier: String?, label: String?, text: String?) -> ElementSelectors {
        let viewPath = identifier != nil ? "/SwiftUI[@id='\(identifier!)']" : "/SwiftUI"
        let hierarchyPath = "/SwiftUI/View"
        let indexedPath = "/SwiftUI[0]/View[0]"
        
        var attributes: [String: String] = [:]
        attributes["framework"] = "SwiftUI"
        
        return ElementSelectors(
            accessibilityIdentifier: identifier,
            accessibilityLabel: label,
            text: text,
            className: "SwiftUIView",
            viewIdPath: viewPath,
            hierarchyPath: hierarchyPath,
            indexedPath: indexedPath,
            bounds: nil,
            attributes: attributes
        )
    }
}

File: demo-app/android/observe-sdk/src/main/java/com/observe/sdk/selectors/SelectorBuilder.kt
package com.observe.sdk.selectors

import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.compose.ui.platform.ComposeView
import org.json.JSONObject

/**
 * Enhanced selector builder for Android native elements
 * 
 * Generates multiple selector strategies with uniqueness validation:
 * 1. Test tag / resource ID (most reliable)
 * 2. Accessibility content description
 * 3. Text content (for buttons, labels)
 * 4. View hierarchy path
 * 5. Indexed path (fallback)
 */
object SelectorBuilder {
    
    /**
     * Element selector information with multiple strategies
     */
    data class ElementSelectors(
        val resourceId: String?,
        val testTag: String?,
        val contentDescription: String?,
        val text: String?,
        val className: String,
        val viewIdPath: String,
        val hierarchyPath: String,
        val indexedPath: String,
        val bounds: Bounds?,
        val attributes: Map<String, Any?>
    ) {
        data class Bounds(
            val x: Int,
            val y: Int,
            val width: Int,
            val height: Int
        )
        
        fun toJson(): String {
            val json = JSONObject()
            json.put("resourceId", resourceId)
            json.put("testTag", testTag)
            json.put("contentDescription", contentDescription)
            json.put("text", text)
            json.put("className", className)
            json.put("viewIdPath", viewIdPath)
            json.put("hierarchyPath", hierarchyPath)
            json.put("indexedPath", indexedPath)
            if (bounds != null) {
                val boundsJson = JSONObject()
                boundsJson.put("x", bounds.x)
                boundsJson.put("y", bounds.y)
                boundsJson.put("width", bounds.width)
                boundsJson.put("height", bounds.height)
                json.put("bounds", boundsJson)
            }
            val attrsJson = JSONObject()
            attributes.forEach { (key, value) ->
                attrsJson.put(key, value)
            }
            json.put("attributes", attrsJson)
            return json.toString()
        }
        
        /**
         * Get primary selector (most reliable available)
         */
        fun getPrimarySelector(): String {
            return when {
                resourceId != null -> "id:$resourceId"
                testTag != null -> "test_tag:$testTag"
                contentDescription != null -> "content_desc:$contentDescription"
                text != null && text.length < 50 -> "text:$text"
                else -> viewIdPath
            }
        }
        
        /**
         * Get all selector strategies in priority order
         */
        fun getAllStrategies(): List<Pair<String, String>> {
            val strategies = mutableListOf<Pair<String, String>>()
            
            resourceId?.let { strategies.add("resourceId" to "id:$it") }
            testTag?.let { strategies.add("testTag" to "test_tag:$it") }
            contentDescription?.let { strategies.add("contentDescription" to "content_desc:$it") }
            text?.let { 
                if (it.length < 50) strategies.add("text" to "text:$it") 
            }
            strategies.add("viewIdPath" to viewIdPath)
            strategies.add("hierarchyPath" to hierarchyPath)
            strategies.add("indexedPath" to indexedPath)
            
            return strategies
        }
    }
    
    /**
     * Build selectors for a native Android view
     */
    fun buildSelectorsForView(view: View, rootView: View): ElementSelectors {
        val resourceId = getResourceId(view)
        val testTag = getTestTag(view)
        val contentDesc = view.contentDescription?.toString()
        val text = extractText(view)
        val className = view.javaClass.simpleName
        
        val viewPath = buildViewPath(view, rootView)
        val hierarchyPath = buildHierarchyPath(view, rootView)
        val indexedPath = buildIndexedPath(view, rootView)
        
        val bounds = ElementSelectors.Bounds(
            x = IntArray(2).also { view.getLocationOnScreen(it) }[0],
            y = IntArray(2).also { view.getLocationOnScreen(it) }[1],
            width = view.width,
            height = view.height
        )
        
        val attributes = extractAttributes(view)
        
        return ElementSelectors(
            resourceId = resourceId,
            testTag = testTag,
            contentDescription = contentDesc,
            text = text,
            className = className,
            viewIdPath = viewPath,
            hierarchyPath = hierarchyPath,
            indexedPath = indexedPath,
            bounds = bounds,
            attributes = attributes
        )
    }
    
    /**
     * Extract resource ID from view
     */
    private fun getResourceId(view: View): String? {
        return try {
            if (view.id != View.NO_ID) {
                view.resources.getResourceEntryName(view.id)
            } else null
        } catch (e: Exception) {
            null
        }
    }
    
    /**
     * Extract test tag from Compose or View
     */
    private fun getTestTag(view: View): String? {
        return try {
            // Try Compose test tag
            view.getTag(androidx.compose.ui.platform.ComposeView::class.java.hashCode())?.toString()
        } catch (e: Exception) {
            null
        }
    }
    
    /**
     * Extract text content from various view types
     */
    private fun extractText(view: View): String? {
        return when (view) {
            is TextView -> view.text?.toString()
            is Button -> view.text?.toString()
            is EditText -> view.hint?.toString() ?: view.text?.toString()
            is CheckBox -> view.text?.toString()
            is RadioButton -> view.text?.toString()
            else -> null
        }
    }
    
    /**
     * Build path using resource IDs where available
     */
    private fun buildViewPath(view: View, rootView: View): String {
        val path = mutableListOf<String>()
        var current: View? = view
        
        while (current != null && current != rootView) {
            val segment = when {
                current.id != View.NO_ID -> {
                    try {
                        val idName = current.resources.getResourceEntryName(current.id)
                        "${current.javaClass.simpleName}[@id='$idName']"
                    } catch (e: Exception) {
                        current.javaClass.simpleName
                    }
                }
                current.contentDescription != null -> {
                    "${current.javaClass.simpleName}[@content-desc='${current.contentDescription}']"
                }
                else -> current.javaClass.simpleName
            }
            path.add(0, segment)
            current = current.parent as? View
        }
        
        return "/" + path.joinToString("/")
    }
    
    /**
     * Build path using class hierarchy
     */
    private fun buildHierarchyPath(view: View, rootView: View): String {
        val path = mutableListOf<String>()
        var current: View? = view
        
        while (current != null && current != rootView) {
            path.add(0, current.javaClass.simpleName)
            current = current.parent as? View
        }
        
        return "/" + path.joinToString("/")
    }
    
    /**
     * Build indexed path (most fragile, but always works)
     */
    private fun buildIndexedPath(view: View, rootView: View): String {
        val path = mutableListOf<String>()
        var current: View? = view
        
        while (current != null && current != rootView) {
            val parent = current.parent
            val index = if (parent is ViewGroup) {
                val childCount = parent.childCount
                (0 until childCount).firstOrNull { i -> parent.getChildAt(i) == current } ?: 0
            } else 0
            
            path.add(0, "${current.javaClass.simpleName}[$index]")
            current = parent as? View
        }
        
        return "/" + path.joinToString("/")
    }
    
    /**
     * Extract all relevant attributes
     */
    private fun extractAttributes(view: View): Map<String, Any?> {
        val attrs = mutableMapOf<String, Any?>()
        
        attrs["clickable"] = view.isClickable
        attrs["enabled"] = view.isEnabled
        attrs["focusable"] = view.isFocusable
        attrs["visible"] = view.visibility == View.VISIBLE
        attrs["alpha"] = view.alpha
        
        when (view) {
            is EditText -> {
                attrs["inputType"] = view.inputType
                attrs["hint"] = view.hint?.toString()
            }
            is CheckBox -> {
                attrs["checked"] = view.isChecked
            }
            is RadioButton -> {
                attrs["checked"] = view.isChecked
            }
            is ImageView -> {
                attrs["hasImage"] = view.drawable != null
            }
        }
        
        return attrs
    }
    
    /**
     * Validate if selector is unique in the view hierarchy
     */
    fun isSelectorUnique(selector: String, rootView: View): Boolean {
        val matchCount = countMatches(selector, rootView)
        return matchCount == 1
    }
    
    /**
     * Count how many views match the given selector
     */
    private fun countMatches(selector: String, rootView: View): Int {
        // Parse selector format
        return when {
            selector.startsWith("id:") -> {
                val id = selector.removePrefix("id:")
                countByResourceId(id, rootView)
            }
            selector.startsWith("test_tag:") -> {
                val tag = selector.removePrefix("test_tag:")
                countByTestTag(tag, rootView)
            }
            selector.startsWith("content_desc:") -> {
                val desc = selector.removePrefix("content_desc:")
                countByContentDescription(desc, rootView)
            }
            selector.startsWith("text:") -> {
                val text = selector.removePrefix("text:")
                countByText(text, rootView)
            }
            selector.startsWith("/") -> {
                // Path-based selector
                countByPath(selector, rootView)
            }
            else -> 0
        }
    }
    
    /**
     * Count views with matching resource ID
     */
    private fun countByResourceId(id: String, rootView: View): Int {
        var count = 0
        traverseHierarchy(rootView) { view ->
            if (getResourceId(view) == id) {
                count++
            }
        }
        return count
    }
    
    /**
     * Count views with matching test tag
     */
    private fun countByTestTag(tag: String, rootView: View): Int {
        var count = 0
        traverseHierarchy(rootView) { view ->
            if (getTestTag(view) == tag) {
                count++
            }
        }
        return count
    }
    
    /**
     * Count views with matching content description
     */
    private fun countByContentDescription(desc: String, rootView: View): Int {
        var count = 0
        traverseHierarchy(rootView) { view ->
            if (view.contentDescription?.toString() == desc) {
                count++
            }
        }
        return count
    }
    
    /**
     * Count views with matching text
     */
    private fun countByText(text: String, rootView: View): Int {
        var count = 0
        traverseHierarchy(rootView) { view ->
            if (extractText(view) == text) {
                count++
            }
        }
        return count
    }
    
    /**
     * Count views matching a path-based selector
     */
    private fun countByPath(path: String, rootView: View): Int {
        var count = 0
        traverseHierarchy(rootView) { view ->
            val viewIdPath = buildViewPath(view, rootView)
            val hierarchyPath = buildHierarchyPath(view, rootView)
            val indexedPath = buildIndexedPath(view, rootView)
            
            if (viewIdPath == path || hierarchyPath == path || indexedPath == path) {
                count++
            }
        }
        return count
    }
    
    /**
     * Traverse view hierarchy and execute callback for each view
     */
    private fun traverseHierarchy(view: View, callback: (View) -> Unit) {
        callback(view)
        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                traverseHierarchy(view.getChildAt(i), callback)
            }
        }
    }

}

